Add code to link Nessus Agent in cloud-init

jsf9k opened the pull request:
## 🗣 Description ##

This pull request makes the necessary changes to link the Nessus Agent to the CDM Tenable/Nessus server via [`cloud-init`](https://cloud-init.io/) at first boot.

## 💭 Motivation and context ##

We must move the configuration of the Nessus Agent link from [cisagov/ansible-role-cdm-nessus-agent](https://github.com/cisagov/ansible-role-cdm-nessus-agent) to here.  This is because the linking command generates a Tenable Tag, which uniquely identifies the instance to the CDM Nessus server.  If we perform this step at AMI build time, then all instances created from that AMI will have identical Tenable Tags.  Nessus cannot handle "duplicated agents" like this, so we need to instead perform the linking via `cloud-init` at first boot.

See also:
* cisagov/ansible-role-cdm-nessus-agent#11
* cisagov/freeipa-server-packer#51
* cisagov/cool-sharedservices-freeipa#36

## 🧪 Testing ##

All `pre-commit` hooks pass.  I have also deployed these changes to our staging COOL environment and verified that they function as intended.

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
* [x] All new and existing tests pass.

jsf9k committed:
Update the instance role so that the instance can read the necessary Nessus Agent parameters from S3

jsf9k committed:
Add cloud-init code to link Nessus Agent

jsf9k committed:
Add missing variable to cloud-init part

The Python code that links the Nessus Agent needs to know the ARN of the role that allows for the reading of SSM Parameter Store parameters.

jsf9k committed:
Convert var.nessus_groups into a comma-delimited list

jsf9k committed:
Allow the account where the instance is being created to assume the SSM Parameter Store role

jsf9k committed:
Add a parameter to the linking script denoting the region where the SSM Parameter Store parameters are stored

jsf9k committed:
Correctly extract parameter value

jsf9k committed:
Correct path to nessuscli executable

jsf9k committed:
Nosec away two Bandit warnings, and add comments explaining why this is safe to do

dav3r commented on cloud-init/link-nessus-agent.py:
```diff
@@ -0,0 +1,79 @@
+# Third-Party Libraries
+import boto3
+
+# Inputs from terraform
```
```suggestion
# Inputs from Terraform
```

dav3r commented on cloud-init/link-nessus-agent.py:
```diff
@@ -0,0 +1,79 @@
+
+# Link the Nessus Agent
+link_cmd = [
+    "/opt/nessus_agent/sbin/nessuscli",
```
Is there any value here in making the path to `nessuscli` a variable?

jsf9k replied:
Yes.  See commit 543f913.

dav3r reviewed (approved):
Nice work! 🦕

jsf9k committed:
Capitalize a word in a comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add a variable for the Nessus Agent install path

jsf9k committed:
Make SSM Parameter Store keys required variables

jsf9k committed:
Update example code

The updates are to make it agree with the changes in #45.

mcdonnnj commented on cloud-init/link-nessus-agent.py:
```diff
@@ -0,0 +1,80 @@
+)
+nessus_key = nessus_key_response["Parameter"]["Value"]
+
+nessus_port_response = ssm_client.get_parameter(
+    Name=NESSUS_PORT_KEY,
+    WithDecryption=True,
+)
+nessus_port = nessus_port_response["Parameter"]["Value"]
```
Nitpicky and I know this is just a simple script, but I think it would be better if the functionality repeated was in a small helper function.

jsf9k replied:
Great catch!  See commit c3c050a.

mcdonnnj commented on README.md:
```diff
@@ -25,12 +28,15 @@ module "ipa0" {
-  subnet_id            = aws_subnet.second_subnet.id
-  tags                 = {
+  domain              = "example.com"
+  hostname            = "ipa.example.com"
```
Just a quick check, should both modules in the example have the same `hostname`?

jsf9k replied:
Great catch!  See commit c3c050a.

mcdonnnj reviewed (approved):
Thanks for your continued work on getting this beast integrated. 💪💪💪 A thought and a question below.

hillaryj reviewed (approved):
🕺🏻

jsf9k committed:
Give unique hostnames to FreeIPA servers in example

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Rework python script

The script now:
* Includes a function to read the parameters from SSM Parameter Store
instead of duplicating code
* Uses typing hints where possible
* Returns the return code from the nessuscli command that performs the
linking

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on cloud-init/link-nessus-agent.py:
```diff
@@ -0,0 +1,86 @@
+
+def get_parameter(ssm_client, parameter_name: str, with_decryption: bool = True) -> str:
+    """Get the value of the specified parameter."""
+    response: dict[str, Any] = ssm_client.get_parameter(
```
Just as a quick mention, using the generics directly for typehinting is technically a Python 3.9+ feature per [PEP-585](https://www.python.org/dev/peps/pep-0585/#backwards-compatibility). Previous versions of Python need to use the the generics in the [typing](https://docs.python.org/3/library/typing.html) library. I don't think it will really impact what's going on in this script, just pointing it out.

jsf9k replied:
I can change it if you think it's necessary.  I didn't catch on to that since I haven't done much (if any) type linting before.

mcdonnnj replied:
If we're likely to be on `3.7`+ we could just slap `from __future__ import annotations` at the top to pull in compatible behavior to play it safe.

jsf9k replied:
See commit da7bd5a.

mcdonnnj reviewed (approved):
Thanks for taking the type to polish up that script a bit! 🏆

jsf9k committed:
Add an import to ensure that type hints don't cause errors

Some of the type hints that are used in this script could cause issues
for Python < 3.9.  Adding this import will ensure that this does not
happen.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>
